### PR TITLE
[2/gcs-mem-kv] Use memory store client when flag is set

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?

This is part of redis removal. In this PR, if `RAY_gcs_storage=memory`, it'll use memory table instead of redis table.
The config setup has to be moved into GcsServer because with the memory table it's transistent.
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by iycheng** [comment]
This should be reviewed after https://github.com/ray-project/ray/pull/20881/files

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 1]
```diff
@@ -64,6 +60,28 @@ void GcsServer::Start() {
...
+    this->main_service_.stop();
+  };
+  ray::rpc::StoredConfig stored_config;
+  stored_config.set_config(config_.config_list);
+  RAY_CHECK_OK(gcs_table_storage_->InternalConfigTable().Put(ray::UniqueID::Nil(),
+                                                             stored_config, on_done));
+  // Run the main service to make the Put in sync mode
+  main_service_.run();
+  // Reset the main service to the initial status
+  main_service_.restart();
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_server.cc:83]: Why do we need a restart()?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 2]
```diff
@@ -64,6 +60,28 @@ void GcsServer::Start() {
...
+  // Init internal config
+  auto on_done = [this](const ray::Status &status) {
+    RAY_CHECK(status.ok()) << "Failed to put internal config";
+    this->main_service_.stop();
+  };
+  ray::rpc::StoredConfig stored_config;
+  stored_config.set_config(config_.config_list);
+  RAY_CHECK_OK(gcs_table_storage_->InternalConfigTable().Put(ray::UniqueID::Nil(),
+                                                             stored_config, on_done));
+  // Run the main service to make the Put in sync mode
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_server.cc:80]: What does this comment mean?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 3]
```diff
@@ -64,6 +60,28 @@ void GcsServer::Start() {
...
 
+  // Init gcs table storage.
+  if (RayConfig::instance().gcs_storage() == "redis") {
+    gcs_table_storage_ = std::make_shared<gcs::RedisGcsTableStorage>(redis_client_);
+  } else if (RayConfig::instance().gcs_storage() == "memory") {
+    gcs_table_storage_ = std::make_shared<InMemoryGcsTableStorage>(main_service_);
+  } else {
+    RAY_LOG(FATAL) << "Unsupported gcs storage: " << RayConfig::instance().gcs_storage();
+  }
+  // Init internal config
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_server.cc:71]:
```suggestion
  // Init internal config.
```

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 1 above)
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_server.cc:83]: That's because of signal catching in the [main function](https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_server_main.cc#L124-L137). 
If not restart, when we hit that part, it'll call the handler immediately

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_server.cc:80]: `  RAY_CHECK_OK(gcs_table_storage_->InternalConfigTable().Put(ray::UniqueID::Nil(),                                                             stored_config, on_done));`
is an async function, so we are not sure when the function finished.

`on_done` will be called in main_service. so we need main_service_.run() to block here and once it finished, it'll resume.

previously it's done in [main function ](https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_server_main.cc#L58-L84) with promise and a separate thread which I feel is super wired.

Comment (mwtian) [reply src/ray/gcs/gcs_server/gcs_server.cc:77]: Maybe rephrase this comment as `Run async operations from InternalConfigTable().Put() above inline.`?

**Review by mwtian — COMMENTED** [review]
Looks good in general. A few comments.

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 4]
```diff
@@ -406,6 +425,7 @@ void GcsServer::InitStatsHandler() {
 
 void GcsServer::InitKVManager() {
   std::unique_ptr<InternalKVInterface> instance;
+  // TODO (yic): Use a factory with configs
```
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_server.cc:428]: I mean the way to construct redis/memory kv should be put into a factory and we use config struct to control this.

**Comment by iycheng** [comment]
@ericl @mwtian  any other comments?

File: src/ray/gcs/gcs_server/gcs_server.h [hunk 5]
```diff
@@ -48,6 +48,8 @@ struct GcsServerConfig {
   bool grpc_based_resource_broadcast = false;
   bool grpc_pubsub_enabled = false;
   std::string log_dir;
+  // This includes the config list of raylet.
```
Comment (mwtian) [inline src/ray/gcs/gcs_server/gcs_server.h:51]: Side note: it seems `config_list` severely lacks documentation upstream and downstream, and its purpose (looks like default raylet config) is not intuitive from its name. Hopefully it can be cleaned up in future.

**Review by mwtian — APPROVED** [review]

File: src/ray/gcs/gcs_server/gcs_server.h (hunk 5 above)
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_server.h:51]: Yes, agreed. I'm not sure what's this :( I only try to make the logic the same as before.

**Comment by iycheng** [comment]
test failure looks unreleated
